File: ncarg2d/src/libncarg/velvct/vvumxy.f
C
C       $Id: vvumxy.f,v 1.8 2008-07-27 00:17:35 haley Exp $
C                                                                      
C                Copyright (C)  2000
C        University Corporation for Atmospheric Research
C                All Rights Reserved
C
C The use of this Software is governed by a License Agreement.
C
C
C-----------------------------------------------------------------------
C
      SUBROUTINE VVUMXY (X,Y,U,V,UVM,XB,YB,XE,YE,IST)
C
C This is a user modifiable routine that allows custom projections of
C the vector space. X and Y give the vector position within the domain
C of the data space. By default, this space is coincident with the
C grid space (i.e. 1 through dimension lengths of the U and V arrays).
C The vector endpoints are output in fractional coordinates (NDC space).
C Note that this is different from the old MXF and MYF routines, which
C output in 'plotter coordinate' space. It also differs from the 
C Conpack routine CPMPXY, which returns values in user space. 
C 
C VVUMXY (Velocity Vector -- User Map X,Y) is called whenever 
C the internal parameter MAP is set to a value other than 0, 1, or 2.
C
C Based on the magnitude and direction of the vector the start and 
C ending points of the vector are returned in NDC space.
C
C Input parameters:
C
C X,Y   -- vector position in the user coordinate system
C U,V   -- vector components from the U,V arrays for this position
C UVM   -- magnitude of the U,V components (supplied for convenience
C          and efficiency - but note that many mappings do not need 
C          this value)
C
C Output parameters:
C
C XB,YB -- starting point of the vector in fractional coordinates
C          (NDC space), before offset based on the value of the
C          vector positioning parameter, VPO.
C XE,YE -- ending point of the vector in fractional coordinates
C          (NDC space), before offset based on the value of the
C          vector positioning parameter, VPO.
C IST   -- status results of the mapping: 0 indicates success -- any
C          non-zero value causes VVECTR to discard the vector at this
C          location.
C
C
C The mapping common block: made available to user mapping routines
C
      COMMON /VVMAP/
     +                IMAP       ,
     +                XVPL       ,XVPR       ,YVPB       ,YVPT       ,
     +                WXMN       ,WXMX       ,WYMN       ,WYMX       ,
     +                XLOV       ,XHIV       ,YLOV       ,YHIV       ,
     +                SXDC       ,SYDC       ,NXCT       ,NYCT       ,
     +                RLEN       ,LNLG       ,INVX       ,INVY       ,
     +                ITRT       ,IWCT       ,FW2W       ,FH2H       ,
     +                DVMN       ,DVMX       ,RBIG       ,IBIG
C
      SAVE /VVMAP/
C
C Math constants
C
      PARAMETER (PDTOR  = 0.017453292519943,
     +           PRTOD  = 57.2957795130823,
     +           P1XPI  = 3.14159265358979,
     +           P2XPI  = 6.28318530717959,
     +           P1D2PI = 1.57079632679489,
     +           P5D2PI = 7.85398163397448) 
C
C --------------------------------------------------------------------
C
C Description of VVMAP contents:
C
C IMAP                - value of the internal parameter 'MAP'
C XVPL,XVPR,YVPB,YVPT - the currently set viewport values. (GETSET
C                       arguments 1, 2, 3, and 4)
C WXMN,WXMX,WYMN,WYMX - the min and max boundaries of user coordinate
C                       space, (usually but not always equivalent to
C                       window coordinates). WXMN and WYMN are true
C                       minimum values even one or both axes is 
C                       inverted. (i.e. WXMN = MIN of GETSET args 5,6;
C                       WYMN = MIN of GETSET args 7,8)
C XLOV,XHIV,YLOV,YHIV - min and max boundaries of the data space, by
C                       default equivalent to the array grid space.
C                       XLOV and YLOV are not necessarily less than 
C                       XHIV and YHIV respectively.
C SXDC,SYDC           - Scaling factors for converting vector component
C                       values into lengths in NDC space.
C NXCT,NYCT           - Length of each dimension of the U and V 
C                       component arrays.
C RLEN                - Length of the maximum vector in user 
C                       coordinates.
C LNLG                - The linear/log mode (GETSET argument 9)
C INVX,INVY           - User coordinates inversion flags: 
C                       0 - not inverted, 1 - inverted
C ITRT                - value of the internal parameter TRT
C IWCT                - not currently used
C FW2W,FH2H           - scale factors for converting from fraction of
C                       viewport width/height to NDC width/height 
C DVMN,DVMX           - min/max vector lengths in NDC
C RBIG,IBIG           - machine dependent maximum REAL/INTEGER values
C
C --------------------------------------------------------------------
C User-defined common block...
C
C --------------------------------------------------------------------
C
C     IF (IMAP .EQ. =user defined mapping code=) THEN
C
C User defined mapping code....
C
C     ELSE
C
C Default mapping:
C
C Note that the default version of VVUMXY included with
C NCAR Graphics 3.2 incorrectly fails to convert the input X and Y
C location into fractional coordinates (NDC space).
C
         IF (X.LT.WXMN .OR. X.GT.WXMX .OR. 
     +        Y.LT.WYMN .OR. Y.GT.WYMX) THEN
            IST = -1
            RETURN
         END IF
         XB=CUFX(X)
         YB=CUFY(Y)
         XE=XB+U*SXDC
         YE=YB+V*SYDC

C     END IF
C
C Done.
C
      RETURN
C
      END

File: ncarg2d/man/vectors/vvumxy.m
.TH VVUMXY 3NCARG "April 1993" UNIX "NCAR GRAPHICS"
.na
.nh
.SH NAME
VVUMXY - 
The user may modify this routine to define a custom mapping
of vectors from a data coordinate system aligned with the
natural boundaries of the vector field to the uniform
normalized device coordinate (NDC) system suitable for
generating a plot on an output device. It has same
parameters as the internal Vectors routine, VVMPXY, used
for the predefined mappings employed when the MAP parameter
has a value between 0 and 2.
.SH SYNOPSIS
CALL VVUMXY (X,Y,U,V,UVM,XB,YB,XE,YE,IST) 
.SH DESCRIPTION 
.IP X 12
(REAL, input) Location of the vector along the first
dimensional axis in the data coordinate system. When MAP is
0, this is the X Axis. If MAP is 1, it is the longitudinal
axis, and if MAP is 2, it is the radial axis. For other
values of MAP, those that cause VVUMXY to be invoked, the
interpretation is up to the author of the mapping routine.
.IP Y 12
(REAL, input) Location of the vector along the second
dimensional axis in the data coordinate system. When MAP is
0, this is the Y Axis. If MAP is 1, it is the latitudinal
axis, and if MAP is 2, it is the angular axis. For other
values of MAP, those that cause VVUMXY to be invoked, the
interpretation is up to the author of the mapping routine.
.IP U 12
(REAL, input) U component of the vector. If TRT is set to
1, the direction of the U component is tangent to the
direction of the first dimensional axis in the data
coordinate system at the location of the vector. If TRT is
set to 0, and MAP has a value of 0 or 2, the direction of
the U component is parallel to the horizontal (X axis) in
NDC space.
.IP V 12
(REAL, input) V component of the vector. If TRT is set to
1, the direction of the V component is normal to the
direction of the first dimensional axis in the data
coordinate system at the location of the vector. If TRT is
set to 0, and MAP has a value of 0 or 2, the direction of
the V component is parallel to the vertical (Y axis) in NDC
space.
.IP UVM 12
(REAL, input) Magnitude of the U and V components,
SQRT(U*U+V*V). Although this value could be calculated
within the routine, it is more efficient for the calling
routine to supply the value as an argument, since it is
needed for other purposes at a higher level.
.IP XB 12
(REAL, output) Location of the vector starting point
along the horizontal (X axis) in NDC space, before
adjustment based on the value of the vector positioning
parameter, VPO.
.IP YB 12
(REAL, output) Location of the vector starting point
along the vertical (Y axis) in NDC space, before adjustment
based on the value of the vector positioning parameter, VPO
.IP XE 12
(REAL, output) Location of the vector ending point along
the horizontal (X axis) in NDC space, before adjustment
based on the value of the vector positioning parameter, VPO
.IP YE 12
(REAL, output) Location of the vector ending point along
the vertical (Y axis) in NDC space before adjustment based
on the value of the vector positioning parameter, VPO
.IP IST 12
(REAL, output) Status of the vector mapping operation:
0 indicates success, negative values indicate that the
mapping failed; positive values are reserved and should not
be used by the implementor of a mapping routine.
.SH USAGE
The user does not call VVUMXY. Vectors calls it only when the
parameter MAP has a value other than 0, 1, or 2, the mappings handled
by Vectors internally. Note that unlike other user-modifiable mapping
routines in NCAR Graphics, such as CPMPXY, that map a single point
into the user coordinate system, this routine returns two points,
representing both ends of the vector, scaled for magnitude, in the
normalized device coordinate (NDC) system. The NDC system is used for
output because, as a coordinate system guaranteed to be rectangular
and uniform, it serves as a convenient reference system to help map
both vector magnitude and direction correctly. The term uniform, as
used in this discussion, means that an arbitrary numerical increment
along either the X or Y axis has the same length given any offset from
the coordinate system origin. The user coordinate system does not
qualify, because it may be log-scaled, or the X units may have a
different size from the Y units.
.sp
In order to implement a custom mapping, you must pick a unique mapping
code (a positive integer greater than 2), and then modify VVUMXY to
recognize and respond to the chosen code. In the standard distribution
of NCAR Graphics, this routine resides in the file, \'vvumxy.f\'.
VVUMXY has access to a common block called VVMAP that contains a
number of variables used to record the current transformation state.
In order to accommodate a variety of mapping implementations, VVMAP
provides more information than normally required. Consider the values
stored in VVMAP as strictly read-only.  One essential member of this
common block is IMAP, which contains the value currently assigned to
the MAP parameter.
.sp
When implementing a non-linear mapping, an iterative differential
technique will most likely be required. Look at the routine, VVMPXY,
in \'vvmpxy.f\', which handles the pre-defined mappings, for examples
of the method. Both the default transformation (MAP set to 0), in
order to account for possible log scaling of the user coordinate axes,
and also the Ezmap projection (MAP set to 1) use such a technique.
Basically the idea is that the vector components must be
proportionally reduced in size enough that an effectively
"instantaneous" angle can be calculated, although they must not become
so small that the calculation is adversely affected by the floating
point precision available for the machine. Additionally, checks must
be put in place to prevent the increment from stepping off the edge of
the coordinate system space. The pre-defined mappings step in the
opposite direction to find the angle whenever an increment in the
original direction would fall off the edge.
.SH ACCESS
To use VVUMXY, load the NCAR Graphics libraries ncarg, ncarg_gks,
and ncarg_c, preferably in that order.  
.SH SEE ALSO
Online:
vectors,
vectors_params,
vvectr,
vvgetc,
vvgeti,
vvgetr,
vvinit,
vvrset,
vvsetc,
vvseti,
vvsetr,
vvudmv,
ncarg_cbind.
.sp
Hardcopy:
NCAR Graphics Fundamentals, UNIX Version
.SH COPYRIGHT
Copyright (C) 1987-2009
.br
University Corporation for Atmospheric Research
.br
The use of this Software is governed by a License Agreement.
